examples/gadgets/simple/trace_open/main.go

```diff
@@ -0,0 +1,120 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	// Operators provide functionality like enrichment and filtering.
+	// Check the operators documentation to learn more: <TODO>
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+
+	// The oci-handler is responsible for handling OCI images
+	ocihandler "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/oci-handler"
+
+	// The ebpf operator handles all eBPF objects: loading eBPF programs,
+	// attaching them, reading maps, etc.
+	_ "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ebpf"
+
+	// The simple operator is a wrapper to make it easier to implement operators
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/simple"
+
+	// Datasources provide access to the data produced by gadgets and operators.
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+
+	// The json formatter is used to format the data as json.
+	igjson "github.com/inspektor-gadget/inspektor-gadget/pkg/datasource/formatters/json"
+
+	// The gadget context is the glue that connects all components together.
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+
+	// The runtime is the piece that will run our gadget.
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime/local"
+)
+
+func do() error {
+	// First of all, we need to define a (simple) operator that we'll use to subscribe
+	// to events. Check the operator documentation to learn more about them.
+	// TODO: documentation
+	// Our operator should be the last of the chain to print information after
+	// all operators have been run.
+	const opPriority = 50000
+	myOperator := simple.New("myHandler", simple.OnInit(func(gadgetCtx operators.GadgetContext) error {
+		// Subscribe to all datasources and print their output as json the terminal
+		// Check the datasources documentation for more information
+		// TODO: link to documentation
+		for _, d := range gadgetCtx.GetDataSources() {
+			jsonFormatter, _ := igjson.New(d,
+				// Show all fields
+				igjson.WithShowAll(true),
+
+				// Print json in a pretty format
+				igjson.WithPretty(true, "  "),
+			)
+
+			d.Subscribe(func(source datasource.DataSource, data datasource.Data) error {
+				jsonOutput := jsonFormatter.Marshal(data)
+				fmt.Printf("%s\n", jsonOutput)
+				return nil
+			}, opPriority)
+		}
+		return nil
+	}))
```

Comment:
Now that I'm working on [the combiner operator](https://github.com/inspektor-gadget/inspektor-gadget/issues/1924#issuecomment-2127209677), I noticed that by subscribing to a data source in the `OnInit` phase there's a risk that a new operator registers a new data source afte us and we miss it. Of course, you can rely on the priority at the moment you add your operator, but what about new operator created later on by other people?

One solution is to never call `GetDataSources` on the `OnInit` phase as other data sources could be registered in operators with lower priority and instead, do it in the `OnStart` phase where all operators are supposed to have already registered their data sources. But again, there's no guarantee.

We need to describe in the documentation to use high priority for operators registering datasource and maybe also tell that it's a good practice to use the `OnInit` phase is for datasource registration and the `OnStart` phase for subscribing to datasources. WDYT?

cc @flyth @mauriciovasquezbernal

Reply:
I drafted [this](https://gist.github.com/flyth/d98befcdd7a1c6558af9af88012ee89a#operators) earlier which contains some information on the lifecycle hooks.

Basically:
* OnInit() / Instantiate(): Add & manipulate DataSources - priority of operator handles initialization order; this is what GetGadgetInfo() returns
* PreStart() (optional): Subscribe to DataSources (to prevent missing some data) - priority determines handling order
* Start(): Emit data, Subscriptions will be evaluated
* Stop(): Stop everything

Regarding subscription priorities: I think filtering should be done before sorting on the server side. Combining should be done (on the client side only) _before_ sorting but after filtering. So sort/filter would run on both client+server.

I hope this will work for 90% of the use cases 😄

Reply:
I think I messed up the examples because there are two priorities:
- operator priority: defines the order on which operators are initialized, started and stopped. 
- subscription priority: order in which subscriptions are called

Those two are independent, an operator with a low priority (is initialized last), could call subscribe with a high priority (its callback will be called first. 

@flyth what's the design decision to have a priority for the operator and not to rely on the order used in WithDataOperators()?

Reply:
> what's the design decision to have a priority for the operator and not to rely on the order used in WithDataOperators()?

Explicitly adding operators using `WithDataOperators()` was an option added later on; most (default) operators register themselves with the registry, so at least for that I think it makes sense to have the priority mentioned inside the operator (although we could make it overridable later on - e.g. the gadget.yaml could specify the order, or we could have a `WithDataOperatorsOrdered()` that would honor the order in which operators are given, overriding their preset priorities).